### PR TITLE
110 EnTT N-body implementation

### LINKED ISSUE #110: Benchmark N-body case study in EnTT
## Reason
To show how our ECS performs against the most popular ECS-library, benchmark the same N-body simulation from #55 in EnTT to see how well it performs.

## Requirements
- [x] An equivalent benchmark to the one in #55 is created in the latest version of EnTT
- [x] The benchmark runs for exactly the same amount of ticks as the one in #55
- [x] The implementation of the N-body system in C++ is as close as possible to the one in Rust

## Dependencies
- [x] #55

### PR DESCRIPTION
**Pull request by EdvinNilsson**
This PR implements the n-body case study using the EnTT ECS-library in C++. The library Eigen3 is used for linear algebra and TBB (Threading Building Blocks) is used for intra-system parallelization.

# Result: Average TPS
| Body Count                 | 100       | 500      | 1000      | 5000       | 10000      |
|----------------------------|-----------|----------|-----------|------------|------------|
| RECS (master)              | 3265.9197 | 709.0834 | 469.5145  | 31.241575  | 8.828015   |
| Unity GameObjects (Mono)   | 244.5293  | 10.02614 | 2.490077  | 0.09792547 | 0.017213   |
| Unity GameObjects (IL2CPP) | 464.7787  | 19.44794 | 4.900632  | 0.1895355  | 0.02723982 |
| Unity ECS (Editor)         | 487.5645  | 411.0888 | 278.5526  | 43.04486   | 11.98565   |
| EnTT (Clang)               | 31897.9   | 3700.28  | 1016.26   | 42.6303    | 10.6667    |
| EnTT (GCC)                 | 28481.9   | 3018.87  | 816.993   | 34.3672    | 8.61475    |
| EnTT (MSVC)               | 14822.1 | 1709.02  | 480.455   | 20.4074    | 5.08776    |

More info about the benchmark results can be found in this [PR](https://github.com/martinjonsson01/recs/pull/113).

Closes #110

### CODE REVIEW

**Review by martinjonsson01 — APPROVED** [review]
LGTM!

Similar questions as for #113:

How difficult do you think it'd be to include some statistical analysis in the benchmarks, like Criterion does? I.e. run the benchmark several times to compute variance and expected value.

Could you add a README explaining how to compile and run the benchmark as a step-by-step?

**Review by Matte02 — APPROVED** [review]
LGTM!

**Comment by linnosterlund** [comment]
LGTM!

**Comment by EdvinNilsson** [comment]
Added build instruction that works with MSVC on Windows.  I also added the benchmark results for EnTT with MSVC.

File: case-studies/n-body/entt/main.cpp [hunk 1]
```diff
@@ -86,33 +84,74 @@ void update(entt::registry& registry) {
...
+          mean += measurement.count();
+      }
+      mean /= measurement_count;
+      
+      double variance = 0;
+      for (auto& measurement : measurements) {
+          auto val = measurement.count();
+          variance += (val - mean) * (val - mean);
+      }
+      auto standard_deviation  = sqrt(variance);
```
Comment (martinjonsson01) [inline case-studies/n-body/entt/main.cpp:135]: Are you sure this computation for variance is correct? I think you might be missing the division by `n - 1`. I'm comparing against [this two-pass algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Two-pass_algorithm).

**Review by martinjonsson01 — APPROVED** [review]
Looks great!

File: case-studies/n-body/entt/main.cpp [hunk 2]
```diff
@@ -85,74 +91,85 @@ void update(entt::registry& registry) {
...
+            registry.emplace<Mass>(entity, 1e6f * radius * radius * radius);
+        }
+
+        for (int i = -1; i < SAMPLE_COUNT; i++) {
+            auto start = std::chrono::high_resolution_clock::now();
+            for (int j = 0; j < ticks_per_sample; j++) {
+                update(registry);
+            }
+            auto stop = std::chrono::high_resolution_clock::now();
+            if (i >= 0) samples[i] = stop - start;
```
Comment (martinjonsson01) [inline case-studies/n-body/entt/main.cpp:125]: is the first `i = -1` iteration a warmup-phase or why does it begin at `-1`?

File: case-studies/n-body/entt/main.cpp [hunk 3]
```diff
@@ -85,74 +91,85 @@ void update(entt::registry& registry) {
...
+    }
 };
 
+struct numpunct : std::numpunct<char> {
+    char do_decimal_point() const override { return ','; }
+};
+
+void save_as_csv(const std::vector<Benchmark>& benchmarks) {
+    std::filesystem::path path = "entt.csv";
+    std::cout << "Writing benchmark results to " << std::filesystem::absolute(path) << "." << std::endl;
```
Comment (martinjonsson01) [inline case-studies/n-body/entt/main.cpp:140]: I'm putting all my `.csv`s at `recs/benchmarking/`. Could you do that as well?

**Review by martinjonsson01 — APPROVED** [review]
LGTM!

File: case-studies/n-body/entt/main.cpp (hunk 2 above)
Comment (EdvinNilsson) [reply case-studies/n-body/entt/main.cpp:125]: Yes, it’s a warmup-phase.

File: benchmarking/bench_n_body.py [hunk 4]
```diff
@@ -112,21 +96,79 @@ def write_results_to_csv(engine_results):
...
-print("done")
+def is_executable(file: Path) -> bool:
+    ext = os.path.splitext(file)[1]
+    return not file.is_dir() and (ext == ".exe" or ext == ".x86_64" or (ext == "" and os.access(file, os.X_OK)))
+
+
+def run_precompiled_benchmarks():
+    precompiled_benchmarks_directory = benchmarking_directory / "precompiled-benchmarks"
+    if not precompiled_benchmarks_directory.exists():
+        print(f"warning: {precompiled_benchmarks_directory} does not exit. skipping precompiled benchmarks.")
```
Comment (martinjonsson01) [inline benchmarking/bench_n_body.py:107]:
```suggestion
        print(f"warning: {precompiled_benchmarks_directory} does not exist. skipping precompiled benchmarks.")
```

File: benchmarking/bench_n_body.py [hunk 5]
```diff
@@ -112,21 +96,79 @@ def write_results_to_csv(engine_results):
...
+        cmd = [binary]
+        if is_unity:
+            cmd += ["-batchmode", "-nographics"]
+
+        proc = subprocess.Popen(cmd, shell=False)
+        proc.wait()
+
+        csv = next(precompiled_benchmark.glob("*.csv"))
+        if not csv.exists():
+            print(f"error: could not find an csv for {binary}.")
```
Comment (martinjonsson01) [inline benchmarking/bench_n_body.py:134]:
```suggestion
            print(f"error: could not find a csv for {binary}.")
```

**Review by martinjonsson01 — APPROVED** [review]
👍👍

**Pull request merged**